sxg_rs/src/link.rs

```diff
@@ -14,6 +15,94 @@ pub(crate) async fn process_link_header(
+
+    let fallback_origin = fallback_url.origin();
+    let directives = RefCell::new(vec![]);
+    stream::iter(preloads)
+        .for_each_concurrent(None, |link| async {
+            let link = link;
+            let srcset = get_param(&link.params, "imagesrcset");
+            let images = match &srcset {
+                Some(srcset) => srcset::parse(srcset).unwrap_or_default(),
+                None => vec![],
+            };
+            // Convert image hrefs from srcset to absolute, for comparison with
+            // allowed_alt_sxgs.
+            let images = images
+                .iter()
+                .filter_map(|url| fallback_url.join(url).ok().map(String::from));
+
+            // Collect allowed-alt-sxg directives for all URLs referred to by this preload.
+            // We will only output the preload if all of them can be found or computed.
+            let mut allow_directives = vec![];
+            let mut all_allow_sxg = true;
+
+            // Iterate over all URLs referred to by this preload.
+            // TODO: Make these concurrent also. (Not critical because imagesrcset is rare.)
+            let urls = once(link.uri.clone()).chain(images);
+            for url in urls {
+                let mut allow_sxg = false;
+                match allowed_alt_sxgs.get(&url) {
+                    Some(allowed_alt_sxg) => {
+                        allow_directives.push(allowed_alt_sxg.clone());
+                        allow_sxg = true;
+                    }
+                    None => {
+                        // Fetch and compute header-integrity only for same-origin preloads.
+                        // Cross-origin preloads are assumed unlikely to be SXG.
+                        if origin_is(&fallback_origin, &url) {
+                            if let Ok(integrity) = header_integrity_fetcher.fetch(&url).await {
+                                allow_directives.push(Link {
+                                    uri: url,
+                                    params: vec![
+                                        ("rel", Some("allowed-alt-sxg".into())),
+                                        ("header-integrity", Some(integrity)),
+                                    ],
+                                });
+                                allow_sxg = true;
+                            }
+                        }
+                    }
+                };
+                if !allow_sxg {
+                    all_allow_sxg = false;
+                    break;
+                }
+            }
+
+            // If all allowed-alt-sxg directives were found, output the preload and allow
+            // directives.
+            if all_allow_sxg {
+                if let Ok(mut directives) = directives.try_borrow_mut() {
+                    directives.push(link.clone());
+                    directives.extend_from_slice(&allow_directives);
+                }
+            }
```

Comment:
Sorry to combine a new feature and a refactoring. The new production code in this file is in these lines. Let me know if you need me to split it up into 2 commits for easier reading.